common/network-types/src/frame.rs

```diff
@@ -0,0 +1,315 @@
+use crate::errors::NetworkTypeError;
+use crate::errors::NetworkTypeError::ReassemblerClosed;
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use futures::{Sink, Stream};
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU16, AtomicU32, Ordering};
+use std::sync::OnceLock;
+use std::task::{Context, Poll};
+
+fn segment(data: &[u8], mtu: u16, frame_id: u16) -> Vec<Segment> {
+    let chunks = data.chunks(mtu as usize);
+    assert!(chunks.len() < u16::MAX as usize, "data too long");
+
+    let seq_len = chunks.len() as u16;
+
+    chunks
+        .enumerate()
+        .map(|(idx, data)| Segment {
+            frame_id,
+            seq_len,
+            seq_idx: idx as u16,
+            data,
+        })
+        .collect()
+}
+
+/// Data frame of arbitrary length.
+/// The frame can be segmented into [segments](Segment) and reassembled back
+/// via [FrameReassembler].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Frame {
+    /// Identifier of this frame.
+    pub frame_id: u16,
+    /// Frame data.
+    pub data: Box<[u8]>,
+}
+
+impl Frame {
+    /// Segments this frame into a list of [segments](Segment) each of maximum sizes `mtu`.
+    pub fn segment(&self, mtu: u16) -> Vec<Segment> {
+        segment(self.data.as_ref(), mtu, self.frame_id)
+    }
+}
+
+/// Represents a frame segment.
+/// Besides the data, a segment carries information about the total number of
+/// segments in the original frame, its index within the frame and
+/// ID of that frame.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Segment<'a> {
+    frame_id: u16,
+    seq_len: u16,
+    seq_idx: u16,
+    data: &'a [u8],
+}
+
+impl<'a> PartialOrd<Segment<'a>> for Segment<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Segment<'_> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match self.frame_id.cmp(&other.frame_id) {
+            std::cmp::Ordering::Equal => self.seq_idx.cmp(&other.seq_idx),
+            cmp => cmp,
+        }
+    }
+}
+
+impl<'a> Segment<'a> {
+    /// Identifies the entire segment sequence using the frame id and the length
+    /// of the sequence.
+    pub fn seq_id(&self) -> u32 {
+        (self.frame_id as u32) << 16 | self.seq_len as u32
+    }
+}
+
+/// Rebuilds the [Frame] from [Segments](Segment).
+#[derive(Debug)]
+struct FrameBuilder<'a> {
+    frame_id: u16,
+    segments: Vec<OnceLock<&'a [u8]>>,
+    missing: AtomicU16,
+}
+
+impl<'a> FrameBuilder<'a> {
+    /// Creates a new builder with the given initial segment.
+    fn new(initial: Segment<'a>) -> Self {
+        Self {
+            frame_id: initial.frame_id,
+            segments: (0..initial.seq_len)
+                .map(|i| {
+                    if i == initial.seq_idx {
+                        OnceLock::from(initial.data)
+                    } else {
+                        OnceLock::new()
```

Comment:
The `OnceLock` has a different purpose here - it makes sure that the `FrameBuilder::put` does not need to take `&mut self`.

And since `OnceLock` is not implemented via actual mutex, but rather atomics, this might be reasonably fast as well.

Reply:
OK, then it's a valid approach.